MapboxMobileEventsTests/MMEAPIClientTests.mm

```diff
@@ -43,6 +47,145 @@ - (void)setupUserAgent;
         apiClient.baseURL should equal([NSURL URLWithString:MMEAPIClientBaseURL]);
     });
     
+    describe(@"- URLSession:didReceiveChallenge:completionHandler:", ^{
+        context(@"when challenge is evaluated by TrustKit", ^{
+            __block bool isMainThread;
+            
+            beforeEach(^{
+                dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+                
+                id<CedarDouble> delegateFake = fake_for(@protocol(NSURLAuthenticationChallengeSender));
+                
+                NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
+                MMENSURLSessionWrapper *sessionWrapper = [[MMENSURLSessionWrapper alloc] init];
+                
+                apiClient.sessionWrapper = sessionWrapper;
+                spy_on(apiClient.sessionWrapper);
+                
+                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, timeoutInNanoseconds), dispatch_get_main_queue(), ^{
+                    dispatch_semaphore_signal(semaphore);
+                });
+                
+                NSURLSession *urlSession = [NSURLSession sessionWithConfiguration:sessionConfig delegate:sessionWrapper delegateQueue:nil];
+                NSURLProtectionSpace *space = [[NSURLProtectionSpace alloc] initWithHost:@"hostname" port:0 protocol:nil realm:nil authenticationMethod:NSURLAuthenticationMethodServerTrust];
+                NSURLAuthenticationChallenge *challenge = [[NSURLAuthenticationChallenge alloc] initWithProtectionSpace:space proposedCredential:nil previousFailureCount:0 failureResponse:nil error:nil sender:((DelegateTestClass *)delegateFake)];
+                
+                [sessionWrapper URLSession:urlSession didReceiveChallenge:challenge completionHandler:^(NSURLSessionAuthChallengeDisposition, NSURLCredential * _Nullable) {
+                    isMainThread = [NSThread isMainThread];
+                    dispatch_semaphore_signal(semaphore);
+                }];
+                
+                while (dispatch_semaphore_wait(semaphore, DISPATCH_TIME_NOW)) {
+                    [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:10]];
+                }
+            });
+            
+            it(@"should be on main queue", ^{
+                isMainThread should be_truthy;
+            });
+            
+            it(@"should receive challenge", ^{
```

Comment:
Hey, sorry I'm late to the party here, but this isn't a particularly valuable assertion, since you called this method yourself in the `beforeEach()`

Instead, it seems that what we care about is that TrustKit's pinning validator gets to handle the challenge ([link](https://github.com/mapbox/mapbox-events-ios/blob/5075b42512ca8675a77b8fd85f0f629eb22c93c3/MapboxMobileEvents/MMENSURLSessionWrapper.m#L73)) and that the inner completion is executed on the main queue. You've got the main queue (thread) part handled, but I'd recommend not spying on the session wrapper, and instead figuring out how to assert that the pinning validator was called as expected (since that's the seam into 3rd party code we care about)

... and if the pinning validator hits the network or does async work of its own, I'd recommend injecting a fake that doesn't introduce that non-determinism into the test suite.

Reply:
Good point: https://github.com/mapbox/mobile-telemetry/issues/291

Reply:
Fair point, these tests stopped being valuable when we changed directions and I failed to notice that. I'll revert...